vllm/v1/engine/llm_engine.py

```diff
@@ -300,155 +290,62 @@ def add_request(
     def abort_request(self, request_id: Union[str, Iterable[str]]) -> None:
         self.scheduler.finish_requests(request_id,
                                        RequestStatus.FINISHED_ABORTED)
-        self._free_request(request_id)
+        # TODO(robertgshaw2): send msg to Detokenizer to free.
+        # (maybe this already happens, need to check)
 
     def get_num_unfinished_requests(self) -> int:
         """Gets the number of unfinished requests."""
-        return len(self.requests)
+        return self.scheduler.get_num_unfinished_requests()
 
     def has_unfinished_requests(self) -> bool:
         """Returns True if there are unfinished requests."""
-        return len(self.requests) > 0
-
-    def step(self) -> List[RequestOutput]:
-        # NOTE(woosuk): This method may return an empty list when the
-        # detokenizer is still processing the outputs. This should not be
-        # considered as the end of the generation process.
-        # FIXME(woosuk): Currently, the step method is inefficient because it
-        # creates RequestOutput objects for all running requests, while they
-        # may not be needed unless the output is streamed to the client.
+        return self.scheduler.has_unfinished_requests()
+
+    def step(self) -> None:
```

Comment:
This technically breaks the API, as `step` no longer returns `RequestOutput`.

Note: this is already how we use the `LLMEngine` in the `MQLLMEngine` with `async_process_outputs`.